scripts/lsp_types.lua

```
@@ -18,19 +19,90 @@ local function tofile(fname, text)
-    'curl',
-    'https://raw.githubusercontent.com/microsoft/lsprotocol/' .. opt.ref .. '/generator/lsp.json',
-    '-o',
-    './lsp.json',
+  return vim.json.decode(res.stdout)
+end
+
+local function gen_methods(protocol)
+  local output =
+    { '--- LSP methods. Generated by lsp_types.lua, keep at end of file.', 'protocol.Methods = {' }
+  local indent = (' '):rep(2)
+
+  vim.iter(protocol.requests):map(function(item)
+    if item.method then
+      local name = item.method:gsub('/', '_', 3)
+      if item.documentation then
+        local document = vim.split(item.documentation, '\n?\n', { trimempty = true })
+        vim.iter(document):map(function(docstring)
+          output[#output + 1] = indent .. '--- ' .. docstring
+        end)
+      end
+      output[#output + 1] = indent .. name .. " = '" .. item.method .. "',"
+    end
+  end)
+  output[#output + 1] = '}'
+  output = vim.list_extend(
+    output,
+    vim.split(
+      [[
+local function freeze(t)
+  return setmetatable({}, {
+    __index = t,
+    __newindex = function()
+      error('cannot modify immutable table')
+    end,
   })
-  local protocol = vim.fn.json_decode(vim.fn.readfile('./lsp.json'))
-  vim.fn.delete('./lsp.json')
-  protocol = protocol or {}
+end
+protocol.Methods = freeze(protocol.Methods)
+
+return protocol
+]],
+      '\n',
+      { trimempty = true }
+    )
+  )
+
+  local fname = './runtime/lua/vim/lsp/protocol.lua'
+  local bufnr = vim.fn.bufadd(fname)
+  vim.fn.bufload(bufnr)
+  vim.api.nvim_set_current_buf(bufnr)
+  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
+  local index = vim.iter(ipairs(lines)):find(function(key, item)
+    return item:find('%-%-%- Lsp methods') and key or nil
+  end)
+  index = index and index - 1 or vim.api.nvim_buf_line_count(bufnr) - 1
+  vim.api.nvim_buf_set_lines(bufnr, index, -1, true, output)
+  vim.cmd.write()
+end
+
+function M.gen(opt)
+  local protocol = read_json(opt)
+  if not protocol then
+    return
```

silently failing is not a good practice. `os.exit(1)`

---

error handle in `read_json`

---

that doesn't set the exit code.